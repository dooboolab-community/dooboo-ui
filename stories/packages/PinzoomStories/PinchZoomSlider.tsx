import {
  Animated,
  Dimensions,
  FlatList,
  Image,
  ImageProps,
  ImageSourcePropType,
} from 'react-native';
import {PinchZoom, PinchZoomRef} from '../../../packages/PinchZoom';
import React, {ReactElement} from 'react';

import styled from '@emotion/native';

const Container = styled.View`
  flex: 1;
  align-self: stretch;
  justify-content: center;
  padding: 10px;
  background-color: #000;
`;

const ImageSliderContainer = styled.View`
  flex: 1;
  align-self: center;
  width: 300px;
  justify-content: center;
  max-width: ${(): number => Dimensions.get('screen').width}px;
  overflow: hidden;
`;

const ItemContainer = styled.View`
  padding: 10px 0;
  max-width: ${(): number => Dimensions.get('screen').width - 20}px;
  min-height: 200px;
`;

const TitleText = styled.Text`
  font-size: 16px;
`;

const ContentText = styled.Text`
  font-size: 14px;
`;

const ImageContainer = styled.View`
  background-color: #ccc;
  width: 100%;
  height: 200px;
  align-items: center;
`;

const images = [
  // IC_LOGO,
  {
    uri: 'https://images.unsplash.com/photo-1519335337423-a3357c2cd12e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2168&q=80',
  },
  {
    uri: 'https://www.housingwire.com/wp-content/uploads/2019/09/Purple-technology-data-internet-3.jpg',
  },
  {
    uri: 'https://images.unsplash.com/photo-1587628736664-fdc50efb57b7?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80',
  },
  {
    uri: 'https://p.bigstockphoto.com/eIdTXLbqQilMs9xbjvcs_bigstock-Aerial-View-Of-Sandy-Beach-Wit-256330393.jpg',
  },
  {uri: 'https://avatars2.githubusercontent.com/u/7970947?v=3&s=460'},
];

interface ImageItemProps {
  source: ImageSourcePropType;
  title: string;
  content: string;
}

function ImageItem({
  source: imageSource,
  title,
  content,
}: ImageItemProps): React.ReactElement {
  const [width, setWidth] = React.useState<string | number>('100%');

  return (
    <ItemContainer>
      <TitleText>{title}</TitleText>
      <ContentText>{content}</ContentText>
      <ImageContainer style={{overflow: 'hidden' /** It may be important! */}}>
        <PinchZoom style={{width, height: 200}}>
          <Image
            style={{width, height: 200, backgroundColor: '#fff'}}
            onLoad={({nativeEvent: {source}}): void => {
              if (source && source.width && source.height)
                setWidth((200 * source.width) / source.height);
            }}
            source={imageSource}
            resizeMode={'contain'}
          />
        </PinchZoom>
      </ImageContainer>
    </ItemContainer>
  );
}

export function ImageList(): React.ReactElement {
  return (
    <Container>
      <FlatList
        data={images}
        style={{flex: 1}}
        keyExtractor={(item): string => item.uri}
        renderItem={({item, index}): React.ReactElement => (
          <ImageItem
            source={item}
            title={`Image ${index + 1}`}
            content="This line is for the description of the image"
          />
        )}
      />
    </Container>
  );
}

function AutoHeightImage(
  props: ImageProps & {style: {width: number}},
): React.ReactElement {
  const {style} = props;
  const [heightPerWidth, setHightPerWidth] = React.useState(0);

  return (
    <Image
      {...props}
      style={[style, {height: heightPerWidth * style.width || style.width}]}
      onLoad={({nativeEvent: {source}}): void => {
        if (source && source.width && source.height)
          setHightPerWidth(source.height / source.width);
      }}
    />
  );
}

const WIDTH = 300;

export const ImageSlider = ({
  imageSources = images,
}: {
  imageSources?: ImageSourcePropType[];
}): ReactElement => {
  const [currentIndex, setCurrentIndex] = React.useState<number>(0);

  const animValues = React.useRef({
    scale: 1,
    x: 0,
    y: 0,
    nextTranslateX: 0,
    prevTranslateX: 0,
  }).current;

  const pinchZoom = React.useRef<PinchZoomRef>(null);
  const nextImageTranslateX = React.useRef(new Animated.Value(0)).current;
  const prevImageTranslateX = React.useRef(new Animated.Value(0)).current;

  const translateOtherImages = React.useCallback(() => {
    animValues.nextTranslateX =
      ((animValues.scale - 1) * WIDTH) / 2 + animValues.x;

    animValues.prevTranslateX =
      ((1 - animValues.scale) * WIDTH) / 2 + animValues.x;

    nextImageTranslateX.setValue(animValues.nextTranslateX);
    prevImageTranslateX.setValue(animValues.prevTranslateX);
  }, [animValues, nextImageTranslateX, prevImageTranslateX]);

  const onRelease = React.useCallback(() => {
    const moveNext = animValues.nextTranslateX < -WIDTH / 2;
    const movePrev = animValues.prevTranslateX > WIDTH / 2;
    const targetTranslate = pinchZoom.current?.animatedValue.translate;

    if (moveNext && currentIndex < imageSources.length - 1 && targetTranslate)
      Animated.timing(targetTranslate, {
        toValue: {x: (-(animValues.scale + 1) / 2) * WIDTH, y: animValues.y},
        useNativeDriver: true,
        duration: 300,
      }).start(() => {
        nextImageTranslateX.setValue(0);
        prevImageTranslateX.setValue(0);
        setCurrentIndex(currentIndex + 1);
        pinchZoom.current?.setValues({scale: 1, translate: {x: 0, y: 0}});
      });
    else if (movePrev && currentIndex > 0 && targetTranslate)
      Animated.timing(targetTranslate, {
        toValue: {x: ((animValues.scale + 1) / 2) * WIDTH, y: animValues.y},
        useNativeDriver: true,
        duration: 300,
      }).start(() => {
        nextImageTranslateX.setValue(0);
        prevImageTranslateX.setValue(0);
        setCurrentIndex(currentIndex - 1);
        pinchZoom.current?.setValues({scale: 1, translate: {x: 0, y: 0}});
      });
    else if (animValues.nextTranslateX < 0 && targetTranslate)
      Animated.timing(targetTranslate, {
        toValue: {x: ((1 - animValues.scale) * WIDTH) / 2, y: animValues.y},
        useNativeDriver: true,
        duration: 300,
      }).start(() => {
        pinchZoom.current?.setValues({
          translate: {
            x: ((1 - animValues.scale) * WIDTH) / 2,
            y: animValues.y,
          },
        });
      });
    else if (animValues.prevTranslateX > 0 && targetTranslate)
      Animated.timing(targetTranslate, {
        toValue: {x: ((animValues.scale - 1) * WIDTH) / 2, y: animValues.y},
        useNativeDriver: true,
        duration: 300,
      }).start(() => {
        pinchZoom.current?.setValues({
          translate: {
            x: ((animValues.scale - 1) * WIDTH) / 2,
            y: animValues.y,
          },
        });
      });
  }, [
    animValues.nextTranslateX,
    animValues.prevTranslateX,
    animValues.scale,
    animValues.y,
    currentIndex,
    imageSources.length,
    nextImageTranslateX,
    prevImageTranslateX,
  ]);

  return (
    <Container style={{backgroundColor: '#000'}}>
      <ImageSliderContainer>
        {currentIndex > 0 ? (
          <PinchZoom
            key={`${currentIndex - 1}`}
            style={{
              position: 'absolute',
              width: WIDTH,
              left: -WIDTH,
              top: 0,
              bottom: 0,
              justifyContent: 'center',
              // @ts-ignore
              transform: [{translateX: prevImageTranslateX}],
            }}>
            <AutoHeightImage
              source={imageSources[currentIndex - 1]}
              style={{width: WIDTH, bottom: 0}}
              resizeMode="contain"
            />
          </PinchZoom>
        ) : null}
        <PinchZoom
          key={`${currentIndex}`}
          ref={pinchZoom}
          onScaleChanged={(value): void => {
            animValues.scale = value;
            translateOtherImages();
          }}
          onTranslateChanged={(value): void => {
            animValues.x = value.x;
            animValues.y = value.y;
            translateOtherImages();
          }}
          onRelease={onRelease}
          fixOverflowAfterRelease={false}
          style={{
            width: WIDTH,
            justifyContent: 'center',
          }}>
          <AutoHeightImage
            source={imageSources[currentIndex]}
            style={{width: WIDTH, bottom: 0}}
            resizeMode="contain"
          />
        </PinchZoom>
        {currentIndex < imageSources.length - 1 ? (
          <PinchZoom
            key={`${currentIndex + 1}`}
            style={{
              width: WIDTH,
              position: 'absolute',
              top: 0,
              bottom: 0,
              justifyContent: 'center',
              left: WIDTH,
              // @ts-ignore
              transform: [{translateX: nextImageTranslateX}],
            }}>
            <AutoHeightImage
              source={imageSources[currentIndex + 1]}
              style={{width: WIDTH}}
              resizeMode="contain"
            />
          </PinchZoom>
        ) : null}
      </ImageSliderContainer>
    </Container>
  );
};

export const Article = (): React.ReactElement => {
  return (
    <PinchZoom style={{width: '100%', backgroundColor: '#eee', padding: 10}}>
      <TitleText>Article Title</TitleText>
      <ContentText style={{minHeight: 100, textAlignVertical: 'center'}}>
        {`This block is for the content of the article. \
PinchZoom works at only Image but also other views. So, let's zoom in with your fingers.
`}
      </ContentText>
    </PinchZoom>
  );
};
