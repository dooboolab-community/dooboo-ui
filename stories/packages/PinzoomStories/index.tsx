import {ContainerDeco} from '../../../storybook/decorators';
import {ImageSlider} from './PinchZoomSlider';
import React from 'react';
import {storiesOf} from '@storybook/react-native';

storiesOf('PinchZoom', module)
  .addDecorator(ContainerDeco)
  .add('Image slider', () => <ImageSlider />);
