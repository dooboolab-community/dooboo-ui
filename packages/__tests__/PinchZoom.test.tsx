import * as React from 'react';

import {ImageSlider} from '../../stories/packages/PinzoomStories/PinchZoomSlider';
import {PanResponderCallbacks} from 'react-native';
import renderer from 'react-test-renderer';

jest.useFakeTimers();

jest.mock('react-native/Libraries/Interaction/PanResponder', () => {
  return {
    create: (
      responderCallback: PanResponderCallbacks,
    ): {panHandlers: {responderCallback: PanResponderCallbacks}} => {
      return {
        panHandlers: {
          responderCallback,
        },
      };
    },
  };
});

describe('PinchZoom of ImageSlider', () => {
  it('should renders without crashing', () => {
    const renderedJSON = renderer.create(<ImageSlider />).toJSON();

    expect(renderedJSON).toMatchSnapshot();
    expect(renderedJSON).toBeTruthy();
  });

  // describe('Interactions', () => {
  //   const rendered: RenderAPI = render(<ImageSlider />);
  //   const pinchZoomContainer = rendered.getAllByTestId('PINCH_ZOOM_CONTAINER')[0];

  //   it('should zoom in by openGesture', () => {
  //     act(() => {
  //       const callBacks = pinchZoomContainer.props.responderCallback;

  //       openGesture.forEach(({ name, nativeEvent, gestureState }) => {
  //         callBacks[name] && callBacks[name]({ nativeEvent }, gestureState);
  //       });
  //     });

  //     const { transform } = pinchZoomContainer.props.style;
  //     const scale = transform.find(({ scale }) => scale != null).scale;
  //     const translateX = transform.find(({ translateX }) => translateX != null).translateX;
  //     const translateY = transform.find(({ translateY }) => translateY != null).translateY;

  //     const { start, end } = getTwoFingerStartEndPositions(openGesture);
  //     const expectedScale = getChangedDistanceRatio({ start, end });

  //     expect(scale).toBeCloseTo(expectedScale, 0.05);

  //     const zoomInPosition = { x: (start.x1 + start.x2) / 2, y: (start.y1 + start.y2) / 2 };

  //     expect((zoomInPosition.x - TEST_CONTAINER_CENTER.x) * scale)
  //       .toBeCloseTo(zoomInPosition.x - TEST_CONTAINER_CENTER.x - translateX);

  //     expect((zoomInPosition.y - TEST_CONTAINER_CENTER.y) * scale)
  //       .toBeCloseTo(zoomInPosition.y - TEST_CONTAINER_CENTER.y - translateY);
  //   });

  //   it('should be moved by moveGesture when it zoomed in', () => {
  //     const { transform } = pinchZoomContainer.props.style;

  //     const prevScale = transform.find(({ scale }) => scale != null).scale;
  //     const prevTranslateX = transform.find(({ translateX }) => translateX != null).translateX;
  //     const prevTranslateY = transform.find(({ translateY }) => translateY != null).translateY;

  //     act(() => {
  //       const callBacks = pinchZoomContainer.props.responderCallback;

  //       moveGesture.forEach(({ name, nativeEvent, gestureState }) => {
  //         callBacks[name] && callBacks[name]({ nativeEvent }, gestureState);
  //       });
  //     });

  //     const { transform: changedTransform } = pinchZoomContainer.props.style;

  //     const scale = changedTransform.find(({ scale }) => scale != null).scale;
  //     const translateX = changedTransform.find(({ translateX }) => translateX != null).translateX;
  //     const translateY = changedTransform.find(({ translateY }) => translateY != null).translateY;

  //     const moveGestureStateList = moveGesture.filter(({ name }) => name === 'onPanResponderMove')
  //       .map(({ gestureState }) => gestureState);

  //     const lastMoveGestureState = moveGestureStateList[moveGestureStateList.length - 1];

  //     expect(scale).toBeCloseTo(prevScale);
  //     expect(translateX).toBeCloseTo(prevTranslateX + lastMoveGestureState.dx);
  //     expect(translateY).toBeCloseTo(prevTranslateY + lastMoveGestureState.dy);
  //   });

  //   it('should zoom out by closeGesture', () => {
  //     act(() => {
  //       const callBacks = pinchZoomContainer.props.responderCallback;

  //       closeGesture.forEach(({ name, nativeEvent, gestureState }) => {
  //         callBacks[name] && callBacks[name]({ nativeEvent }, gestureState);
  //       });
  //     });

  //     const { transform } = pinchZoomContainer.props.style;
  //     const scale = transform.find(({ scale }) => scale != null).scale;
  //     const translateX = transform.find(({ translateX }) => translateX != null).translateX;
  //     const translateY = transform.find(({ translateY }) => translateY != null).translateY;

  //     const openScale = getChangedDistanceRatio(getTwoFingerStartEndPositions(openGesture));
  //     const closeScale = getChangedDistanceRatio(getTwoFingerStartEndPositions(closeGesture));

  //     expect(openScale * closeScale).toBeLessThan(1);

  //     expect(scale).toEqual(1);
  //     expect(translateX).toEqual(0);
  //     expect(translateY).toEqual(0);
  //   });

  //   it('should release onTranslateChanged function if the property changed.', () => {
  //     const callBacks = pinchZoomContainer.props.responderCallback;

  //     act(() => {
  //       callBacks.onPanResponderGrant();

  //       callBacks.onPanResponderMove({
  //         nativeEvent: {
  //           identifier: 0,
  //           locationX: 247,
  //           locationY: 95,
  //           touches: [
  //             'Self',
  //           ],
  //         },
  //       }, { dx: 0, dy: 0 });

  //       callBacks.onPanResponderMove({
  //         nativeEvent: {
  //           identifier: 0,
  //           locationX: 47,
  //           locationY: 95,
  //           touches: [
  //             'Self',
  //           ],
  //         },
  //       }, { dx: -200, dy: 0 });
  //     });

  //     const { transform } = pinchZoomContainer.props.style;
  //     const scale = transform.find(({ scale }) => scale != null).scale;
  //     const translateX = transform.find(({ translateX }) => translateX != null).translateX;
  //     const translateY = transform.find(({ translateY }) => translateY != null).translateY;

  //     expect(scale).toEqual(1);
  //     expect(translateX).toEqual(-200);
  //     expect(translateY).toEqual(0);

  //     act(() => {
  //       // Call onRelease and the pinchzoom move to -300
  //       callBacks.onPanResponderRelease();
  //     });

  //     jest.runAllTimers();
  //     expect(pinchZoomContainer.props.style.left).toEqual(-300);
  //   });
  // });
});
