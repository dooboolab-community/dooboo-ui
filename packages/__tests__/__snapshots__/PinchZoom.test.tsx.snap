// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PinchZoom of ImageSlider should renders without crashing 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
      },
      Object {
        "backgroundColor": "#000",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#000",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "height": 400,
          "justifyContent": "center",
          "overflow": "hidden",
          "width": 300,
        },
        undefined,
      ]
    }
  >
    <View
      forwardedRef={[Function]}
      style={
        Object {
          "justifyContent": "center",
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
            Object {
              "scale": 1,
            },
          ],
          "width": 300,
        }
      }
      testID="PINCH_ZOOM_CONTAINER"
    >
      <Image
        onLoad={[Function]}
        resizeMode="contain"
        source={
          Object {
            "uri": "https://images.unsplash.com/photo-1519335337423-a3357c2cd12e?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2168&q=80",
          }
        }
        style={
          Array [
            Object {
              "bottom": 0,
              "width": 300,
            },
            Object {
              "height": 300,
            },
          ]
        }
      />
    </View>
    <View
      forwardedRef={[Function]}
      style={
        Object {
          "bottom": 0,
          "justifyContent": "center",
          "left": 300,
          "position": "absolute",
          "top": 0,
          "transform": Array [
            Object {
              "translateX": 0,
            },
          ],
          "width": 300,
        }
      }
      testID="PINCH_ZOOM_CONTAINER"
    >
      <Image
        onLoad={[Function]}
        resizeMode="contain"
        source={
          Object {
            "uri": "https://www.housingwire.com/wp-content/uploads/2019/09/Purple-technology-data-internet-3.jpg",
          }
        }
        style={
          Array [
            Object {
              "width": 300,
            },
            Object {
              "height": 300,
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;
